lock-api/src/main/java/com/palantir/lock/watch/LockWatchEventTrackerImpl.java

```diff
@@ -0,0 +1,174 @@
+
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Maps;
+import com.google.common.collect.RangeSet;
+import com.google.common.collect.TreeRangeSet;
+import com.google.common.primitives.Ints;
+import com.palantir.lock.LockDescriptor;
+
+/**
+ * Note on concurrency:
+ *
+ * Multiple threads may simultaneously try to update the tracker, potentially with overlapping updates. At the same
+ * time, it is allowed to return a slightly stale version on calls to currentState, so we do not want to block on
+ * updates. This is implemented as follows.
+ *
+ * 1. Updates are synchronized, and the values of watches, singleLocks, openLocksEvents, lastKnownVersion, and leaderId
+ *    are only changed within the synchronized block. This ensures that two threads are not going to conflict during an
+ *    update, or do unnecessary work.
+ * 2. Furthermore, changes to the values of watches, singleLocks, openLocksEvents, lastKnownVersion are also guarded by
+ *    a write lock. This allows calls to currentState() to return a consistent version of lock watch state even if an
+ *    update is in progress, as long as the write lock is not locked
+ */
+@ThreadSafe
+public class LockWatchEventTrackerImpl implements LockWatchEventTracker {
+    private final ReadWriteLock lock = new ReentrantReadWriteLock();
+    private final AtomicReference<RangeSet<LockDescriptor>> watches = new AtomicReference<>(TreeRangeSet.create());
+    private final AtomicReference<Map<LockDescriptor, LockWatchInfo>> singleLocks = new AtomicReference<>(
+            ImmutableMap.of());
+    private final ConcurrentSkipListSet<UUID> openLocksEvents = new ConcurrentSkipListSet<>();
+    private final ConcurrentMap<Long, VersionedLockWatchState> startTsToLockWatchState = Maps.newConcurrentMap();
+    private volatile OptionalLong lastKnownVersion = OptionalLong.empty();
+    private volatile UUID leaderId = UUID.randomUUID();
+
+    @Override
+    public VersionedLockWatchState currentState() {
+        lock.readLock().lock();
+        try {
+            return new VersionedLockWatchStateImpl(lastKnownVersion, watches.get(), singleLocks.get(), leaderId);
+        } finally {
+            lock.readLock().unlock();
+        }
+    }
+
+    @Override
+    public synchronized VersionedLockWatchState updateState(LockWatchStateUpdate update) {
+        if (!update.success()) {
+            resetAllState(update);
+            return currentState();
+        }
+
+        if (leaderId != update.leaderId()) {
+            runUpdate(true, update);
+            return currentState();
+        }
+
+        if (!lastKnownVersion.isPresent()) {
+            resetAllState(update);
+            return currentState();
+        }
+
+        long currentTrackedVersion = lastKnownVersion.getAsLong();
+        if (updateNotPossible(update, currentTrackedVersion)) {
+            return currentState();
+        }
+
+        runUpdate(false, update);
+        return currentState();
+    }
```

Comment:
Some comments/explanation probably needed here, I'm having trouble understanding the difference between `resetAllState` vs `runUpdate(true, x)` without staring at what the implementation actually does.

Reply:
Tried to make it more understandable, also changed the case for no known last state, as we can use the update then